New IDAaaS data cache index file  loading

GuiMacielPereira opened the pull request:
### Description of work
Implemented search in data cache for files before defaulting to search in the data archive. This implementation leaves the previous implementation mostly untouched, and only adds a search in the data cache before advancing to the search in the archive. So the order for searching is User Directories > Data Cache > Data archive. This gives priority to local changes that users have made.

#### Summary of work
Added `ISISInstrDataCache.h` and `ISISInstrDataCache.cpp` which is where the implementation resides. Other than that, changed `FileFinder.cpp` to call the search. Added unit tests.

#### Purpose of work
To give access to data files through the data cache on IDAaaS even when the user does not have access to the archive.

Fixes #37013 

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
Added a few comments and print statements to make it clear what is going on. We can remove them during the review stage.

### To test:
Build to test: https://builds.mantidproject.org/job/build_packages_from_branch/792/

1. Log into IDAaaS
2. Create a new conda environment and install build with `mamba install -c mantid/label/idaas_data_cache_debug mantidworkbench`
3. Open workbench
4. Open the file `/data/instrument/ALF/ALF_index.json`. The numbers on the left side correspond to available run numbers
5. In the workbench, change the default instrument to ALF and turn off access to the data archive.
6. Open the Load GUI and type in several run numbers of your choice chosen from the `ALF_index.json` file. You don't need to click `Run`, if the `OutputWorkspace` field gets updated then the file was successfully found. Try this several times and check they all successfully found the file.
7. Try some bad inputs (the error message will be displayed when hovering over the red asterisk):
    - A run number that doesn't exist, for example `99999`
    - Filename with wrong format, eg. `ALF`, `1234ALF`
    - File we don't have access to, eg. `MAR27340`

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

GuiMacielPereira committed:
Simple barebones for searching new index files

- Created a very simplified implementation and a unit test

GuiMacielPereira committed:
Added setUp and tearDown to tests.

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Changed implementation from DataHandling to API

- I made the change because header files from DataHandling could not be imported into files in the API, due to DataHandling being missing from target_link_libraries in the CMakeLists.txt API file. Trying to fix this causes a circular dependency.

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Rough sketch of implementation after looking for data archive

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Add cache search to the file finder.

- Added section to look for files in hard coded Linux path
- Changed search function to use relative paths
- Corrected unit test

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Changing function to handle expanded instr dirs

- The name of the directories in the cache uses the extended version i.e. 'MARI' instead of 'MAR', changed the function to account for this
- Changed split into instr and run to handle edge case of 'SANS2D' (digit in instrument name)

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Add functunality to getInstrument() function

- Function getInstrument() in the ConfigService is used to retrieve the instrument info from a string argument.
- The argument of the function is either the name or short name of the instrument
- Added option to pass the name or short name + delimiter and retrieve the correct instrument

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added test for delimiter in filename

- Added unit test for case when there is a delimiter between instrument name and run number
- Also checks that short name `PG3` gets properly expanded to `POWGEN`

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added handling of bad inputs

- Added some fail proof checks
- Added unit tests for these cases

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added unit tests to FileFinder

- Added a test class that creates a dummy directory with the same structure as the data cache and performs unit tests on the path outputs
- Introduced safety checks in ISISInstrDataCache to cover for edge cases
- Added the data cache path to Mantid properties file

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added exception for missing read permissions

- When user does not have access permissions to a folder, it will print an error informing the user

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added unit test for folder without permissions

GuiMacielPereira commented on Framework/Kernel/src/FacilityInfo.cpp:
```diff
@@ -187,6 +187,19 @@ const InstrumentInfo &FacilityInfo::instrument(std::string iName) const {
                               [&iName](const auto &inst) { return boost::iequals(inst.shortName(), iName); });
   }
 
+  // Cover edge case where delimiter is present
+  if (instrument == m_instruments.cend()) {
+    instrument = std::find_if(m_instruments.cbegin(), m_instruments.cend(), [&iName](const auto &inst) {
+      return boost::iequals(inst.name() + inst.delimiter(), iName);
+    });
+  }
+
+  if (instrument == m_instruments.cend()) {
+    instrument = std::find_if(m_instruments.cbegin(), m_instruments.cend(), [&iName](const auto &inst) {
+      return boost::iequals(inst.shortName() + inst.delimiter(), iName);
+    });
+  }
+
```
I have a strong suspicion this is not needed, because I remember that FileFinder already has its own getInstrument function that includes this check for the delimiter. So I can remove this change and use the FileFinder's function instead, which also makes more sense.

GuiMacielPereira replied:
Took it out without changing functunality

GuiMacielPereira committed:
Removed unecessary changes

- Deleted unecessary changes that I had made
- Replaced instrument finder by one that already accounts for delimiters and other edge cases
- Removed unecessary notice()

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added documentation for new mantid property

- Added datacachesearch.directory to properties file documentation

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Fixed cppcheck warnings

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Cleaned exceptions catching

- Organized catching of exceptions for bad inputs
- Added check for when user does not have permission to open directory
- Changed logging from notice level to debug level

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Fixed failing unit test

- Made a quick fix for the getInstrument function, to make it throw the exception instead of using a default value
- Not proud of this fix, but when I tried to take out the exception handling out of getInstrument, a few other things were breaking

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/inc/MantidAPI/FileFinder.h:
```diff
@@ -49,10 +50,11 @@ class MANTID_API_DLL FileFinderImpl {
   std::vector<std::string> findRuns(const std::string &hintstr, const std::vector<std::string> &exts = {},
                                     const bool useExtsOnly = false) const;
   /// DO NOT USE! MADE PUBLIC FOR TESTING ONLY.
-  const Kernel::InstrumentInfo getInstrument(const std::string &hint) const;
+  const Kernel::InstrumentInfo getInstrument(const std::string &hint, const bool &throwOnThrow = false) const;
```
They do have a big warning saying not to use this method, but I believe they meant for it not to be used instead of `ConfigService::Instance().getInstrument()`, which is the general one. I am using this method in `ISISInstrDataCache.cpp` which is only used in `FileFinder.cpp` so I think it's okay for me to use it?

GuiMacielPereira commented on Framework/API/src/ISISInstrDataCache.cpp:
```diff
@@ -0,0 +1,69 @@
+#include "MantidKernel/InstrumentInfo.h"
+#include "MantidKernel/Logger.h"
+#include <algorithm>
+#include <boost/lexical_cast.hpp>
```
Forgot to clean some of these imports from earlier implementations

jhaigh0 commented on Framework/API/src/ISISInstrDataCache.cpp:
```diff
@@ -0,0 +1,69 @@
+
+  // Get instrument full name from short name
+  std::transform(fileName.begin(), fileName.end(), fileName.begin(), toupper);
+  std::string instrName = fileName.substr(0, nChars);
+  try {
+    auto instrInfo = FileFinder::Instance().getInstrument(instrName, true);
+    instrName = instrInfo.name();
```
I don't think `instrName` needs to be assigned on line 42 if it will always get overwritten on line 45?

jhaigh0 commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -741,8 +787,24 @@ const API::Result<std::string> FileFinderImpl::getPath(const std::vector<IArchiv
+    }
+    errors += cacheFilePath.errors();
+
+  } else {
+    errors += "Could not find data cache directory: " + cachePathToSearch.string();
+  }
```
Will this always happen when not on IDAaaS? If so will it be annoying / confusing?

GuiMacielPereira replied:
No, if the file is then found on the archive then the errors string is never displayed. Otherwise, if the file is neither at the data cache nor at the archive the errors string contains both the data cache and archive errors, and will be displayed in the tooltip of the `Load` function.

jhaigh0 replied:
Ah, so the errors only appear if the file isn't found anywhere

GuiMacielPereira replied:
Yes, exactly, but even then it doesn't show as an error message, it goes into the tooltip

jhaigh0 reviewed (changes requested):
Looking pretty good already, some things to improve.
Other than the comments I've added, I think we can call these files / classes `ISISInstrumentDataCache` etc. We don't need to cut down on file name length (sorry this is a bit of a pain).
Also, I think `getFileParentDirPath` could be broken up into smaller functions, I've added one suggestion but there might be other places you could do this.

GuiMacielPereira committed:
Update Framework/API/inc/MantidAPI/ISISInstrDataCache.h

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

GuiMacielPereira committed:
Change argument flag in getInstrument

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Updated doc string

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Removed unused includes

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Removed another unused include

GuiMacielPereira committed:
Made error message clearer

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Changed to const auto types

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Split big function into smaller functions

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>
Co-authored-by: Caila Finn  <caila.finn@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/test/FileFinderTest.h:
```diff
@@ -465,6 +469,11 @@ class FileFinderTest : public CxxTest::TestSuite {
 
 class FileFinderTestPerformance : public CxxTest::TestSuite {
 public:
+  void setUp() override {
+    // TODO: temporary fix
+    Mantid::Kernel::ConfigService::Instance().setString("datacachesearch.directory", "");
+    Mantid::Kernel::ConfigService::Instance().setLogLevel("debug");
+  }
```
No it wasn't :)

GuiMacielPereira committed:
Removed usused old setUp()

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added comment

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Reformatted strings and removed unecessary ;

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/test/ISISInstrDataCacheTest.h:
```diff
@@ -0,0 +1,111 @@
+        ofstrm.close();
+      }
+    }
+  };
+
+  void tearDown() override { std::filesystem::remove_all(m_dataCacheDir); };
```
Oh my goodness I didn't even realise that my LSP didn't warn about this! I went through all of the files I changed and corrected other places where I had done this as well.

GuiMacielPereira committed:
Included comment for clarification

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Renamed classes and files to more explicit descriptions

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -185,7 +191,7 @@ const Kernel::InstrumentInfo FileFinderImpl::getInstrument(const string &hint) c
  */
 std::pair<std::string, std::string> FileFinderImpl::toInstrumentAndNumber(const std::string &hint) const {
-  // g_log.debug() << "toInstrumentAndNumber(" << hint << ")\n";
+  g_log.debug() << "toInstrumentAndNumber(" << hint << ")\n";
```
Uhm, I uncommented it because when I was trying to understand what the code did it was useful to have this output to the console. But since it was commented out before I can also take it out?

jhaigh0 replied:
I think its fine to leave in that case 👍

GuiMacielPereira committed:
Fixed cppcheck and added unit tests

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/test/FileFinderTest.h:
```diff
@@ -611,3 +623,113 @@ class FileFinderTestPerformance : public CxxTest::TestSuite {
+                     m_dataCacheDir + "/MERLIN/SUBDIR1/SUBDIR2/MER40871.nxs");
+  }
+
+  void testZeroPadding() {
+    TS_ASSERT_EQUALS(FileFinder::Instance().findRun({"WISH39495"}, {".raw", ".nxs", ".s01"}).result(),
+                     m_dataCacheDir + "/WISH/SUBDIR1/SUBDIR2/WISH00039495.s01");
+    TS_ASSERT_EQUALS(FileFinder::Instance().findRun({"LOQ106084"}, {".raw", ".nxs"}).result(),
+                     m_dataCacheDir + "/LOQ/SUBDIR1/SUBDIR2/LOQ00106084.nxs");
+    TS_ASSERT_EQUALS(FileFinder::Instance().findRun({"ZOOM4656"}, {".RAW"}).result(),
+                     m_dataCacheDir + "/ZOOM/SUBDIR1/SUBDIR2/ZOOM00004656.RAW");
+  }
```
When I call the constructor with `ZOOM4656.RAW` it produces a filename with the padded zeros before it creates the file. This is done in line https://github.com/mantidproject/mantid/pull/37187/files#diff-4a4d9fa3c8e1dfda41c071f353ec49736154c2f5f542ff6a4783d0642f6dd859R664

GuiMacielPereira committed:
Added release note (+ section for Data Handling)

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson reviewed (dismissed):
There is an intermittent issue with json parsing on IDAaaS that we are investigating.

sf1919 commented:
Also be aware there seems to be an MSBuild warning - https://builds.mantidproject.org/job/pull_requests-conda-windows/5886/msbuild/

GuiMacielPereira committed:
Cleaned error handling for permission denied

- Check for permission denied is now done inside the loop that directly tries to access files
- Added check for non existing parent directory
- Improved error message when file is not found whithin parent directory

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Added catch for error from parsing json

- If for some reason the index file on IDAaaS is currupted (eg. being updated) then the error is caught
- Allows for the search to continue on the archive

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira committed:
Fixed include error + cppcheck supressions

Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

GuiMacielPereira commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -645,6 +652,53 @@ std::vector<std::string> FileFinderImpl::findRuns(const std::string &hintstr, co
+
+    std::string parentDirPath;
+
+    try {
+      parentDirPath = dataCache.getFileParentDirectoryPath(filename);
+
+    } catch (const std::invalid_argument &e) {
```
Uhm I find the option without the empty line harder to read :/

jhaigh0 replied:
No problem, I thought it was a mistake because a recent commit changed it to this.

GuiMacielPereira committed:
Corrected unit test

robertapplin commented on Framework/API/inc/MantidAPI/ISISInstrumentDataCache.h:
```diff
@@ -0,0 +1,30 @@
+class MANTID_API_DLL ISISInstrumentDataCache {
+public:
+  ISISInstrumentDataCache(const std::string &path) : m_dataCachePath(path) {}
+  std::string getFileParentDirectoryPath(std::string filename);
+
+private:
+  std::pair<std::string, std::string> validateInstrumentAndNumber(std::string filename);
+  std::pair<std::string, std::string> splitIntoInstrumentAndNumber(std::string filename);
```
I think all these functions can also be `const`

GuiMacielPereira replied:
The reason I am not passing these as `const &` is because I am using the method `FileFinder::Instance()::extractAllowedSuffix()` on line https://github.com/mantidproject/mantid/pull/37187/files#diff-ab9dd41f9e44e9c2d5c363cc67efbd0e441a8322ac636636a86f68971f1cf935R47, which does take a reference and might change the string. I tried changing this method in the code before but it was causing other unexpected behaviours so I decided to instead pass the argument as a copy to keep this PR as localized as possible (to avoid changing older parts of the code). In the other function, `splitIntoInstrumentAndNumber`, I am using a `std::transform` inside the function, which alters the string argument. I could change this by passing the argument as `const &` and then creating an internal copy, and finally performing the transform. Do you think that would be preferred?

GuiMacielPereira replied:
I can add const to these methods, yes :+1:

robertapplin replied:
Ah I see, thanks for explaining. I think it would be preferred to create an internal copy in each of these methods

GuiMacielPereira replied:
I see, just did that :+1:

robertapplin reviewed (changes requested):
Sorry for holding this up, just have a few suggestions, then happy for this to be merged

GuiMacielPereira committed:
Removed unused includes + added const to methods

GuiMacielPereira commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -132,9 +136,11 @@ std::string FileFinderImpl::extractAllowedSuffix(std::string &userString) const
  * @return This will return the default instrument if it cannot be determined.
  */
-const Kernel::InstrumentInfo FileFinderImpl::getInstrument(const string &hint) const {
+const Kernel::InstrumentInfo FileFinderImpl::getInstrument(const string &hint,
+                                                           const bool &returnDefaultIfNotFound) const {
```
Thank you for this, I wasn't aware of it!

GuiMacielPereira committed:
Pass bool by value

GuiMacielPereira commented on Framework/API/inc/MantidAPI/ISISInstrumentDataCache.h:
```diff
@@ -0,0 +1,30 @@
+// Includes
+//----------------------------------------------------------------------
+#include "MantidAPI/DllConfig.h"
+#include "MantidAPI/IArchiveSearch.h"
```
Actually I had forgotten this include was still in there, it's not needed so I'll remove it entirely

GuiMacielPereira committed:
Fix Cppcheck

GuiMacielPereira committed:
Passed string by const & + added unit test

robertapplin reviewed (commented):
Thanks for the changes! @jhaigh0 if you're happy then I'll enable auto merge
